src/backend/x11/mod.rs

```diff
@@ -0,0 +1,921 @@
+    ) -> Result<X11Surface, X11Error> {
+        let connection = &backend.connection;
+        let window = backend.window();
+
+        // Determine which drm-device the Display is using.
+        let screen = &connection.setup().roots[backend.screen()];
+        let dri3 = connection.dri3_open(screen.root, 0)?.reply()?;
+
+        let drm_device_fd = dri3.device_fd;
+        // Duplicate the drm_device_fd
+        let drm_device_fd: RawFd = fcntl::fcntl(
+            drm_device_fd.as_raw_fd(),
+            fcntl::FcntlArg::F_DUPFD_CLOEXEC(3), // Set to 3 so the fd cannot become stdin, stdout or stderr
+        )
+        .map_err(AllocateBuffersError::from)?;
+
+        let fd_flags =
+            nix::fcntl::fcntl(drm_device_fd.as_raw_fd(), nix::fcntl::F_GETFD).expect("Handle this error");
+        // No need to check if ret == 1 since nix handles that.
+
+        // Enable the close-on-exec flag.
+        nix::fcntl::fcntl(
+            drm_device_fd.as_raw_fd(),
+            nix::fcntl::F_SETFD(
+                nix::fcntl::FdFlag::from_bits_truncate(fd_flags) | nix::fcntl::FdFlag::FD_CLOEXEC,
+            ),
+        )
+        .map_err(AllocateBuffersError::from)?;
+
+        if get_drm_node_type_from_fd(drm_device_fd.as_raw_fd())? != DRM_NODE_RENDER {
+            todo!("Attempt to get the render device by name for the DRM node that isn't a render node")
+        }
+
+        // Finally create a GBMDevice to manage the buffers.
+        let device = gbm::Device::new(drm_device_fd.as_raw_fd()).expect("Failed to create gbm device");
+
+        let size = backend.window().size();
+        // TODO: Dont hardcode format.
+        let current = device
+            .create_buffer_object::<()>(size.w as u32, size.h as u32, format, BufferObjectFlags::empty())
+            .map_err(Into::<AllocateBuffersError>::into)?
+            .export()
+            .map_err(Into::<AllocateBuffersError>::into)?;
+
+        let next = device
+            .create_buffer_object::<()>(size.w as u32, size.h as u32, format, BufferObjectFlags::empty())
+            .map_err(Into::<AllocateBuffersError>::into)?
+            .export()
+            .map_err(Into::<AllocateBuffersError>::into)?;
+
+        Ok(X11Surface {
+            connection: Arc::downgrade(connection),
+            window,
+            device,
+            format,
+            width: size.w,
+            height: size.h,
+            current,
+            next,
+            resize,
+        })
+    }
+
+    /// Returns a handle to the GBM device used to allocate buffers.
+    pub fn device(&self) -> gbm::Device<RawFd> {
+        self.device.clone()
+    }
+
+    /// Returns the format of the buffers the surface accepts.
+    pub fn format(&self) -> DrmFourcc {
+        self.format
+    }
+
+    /// Returns an RAII scoped object which provides the next buffer.
+    ///
+    /// When the object is dropped, the contents of the buffer are swapped and then presented.
+    // TODO: Error type
```

Comment:
Sorry, but I do not understand this todo. Same for the next function.

Reply:
Forgot to remove these